content/oci/oci.go

```diff
@@ -269,14 +267,35 @@ func (s *Store) SaveIndex() error {
 	defer s.indexLock.Unlock()
 
 	var manifests []ocispec.Descriptor
+	tagged := make(containers.Set[digest.Digest])
 	refMap := s.tagResolver.Map()
+
+	// 1. Add descriptors that are associated with tags
+	// Note: One descriptor can be associated with multiple tags.
+	for ref, desc := range refMap {
+		if ref != desc.Digest.String() {
+			// copy the original annotation map
+			annotations := make(map[string]string, len(desc.Annotations)+1)
+			for k, v := range desc.Annotations {
+				annotations[k] = v
+			}
+			annotations[ocispec.AnnotationRefName] = ref
+			desc.Annotations = annotations
+			manifests = append(manifests, desc)
+			// mark the digest as tagged for deduplication in step 2
+			tagged.Add(desc.Digest)
+		}
+	}
+	// 2. Add descriptors that are not associated with any tag
 	for ref, desc := range refMap {
-		if ref == desc.Digest.String() && desc.Annotations[ocispec.AnnotationRefName] != "" {
-			// skip saving desc if ref is a digest and desc is tagged
-			continue
+		if ref == desc.Digest.String() {
+			// skip tagged ones since they have been added in step 1
+			if !tagged.Contains(desc.Digest) {
+				manifests = append(manifests, desc)
```

Comment:
Just realized that, here we assume desc does not contain the annotation ref name. However, there might be issues if we load the index and save it back. We may need to strip the ref name when loading the index.

Reply:
Fixed.